### ISSUE TITLE
Common shape ops for SparseTensor

### ISSUE THREAD

**Issue by concretevitamin** [tensorflow/tensorflow#1968]
It should be easy to extend/modify `tf.rank()`, `tf.shape()`, and `tf.size()` to have them work on `SparseTensor`s.

Happy to review PRs!

**Comment by AnishShah** [comment]
Hi! 
I'm working on this.

**Comment by AnishShah** [comment]
@concretevitamin I'm having some problem with bazel-build. Not able to build from source. Is there any other way I can build from source?

**Comment by concretevitamin** [comment]
What are the errors?  I trust that the latest 0.8 docs on the website did
not help?

**Comment by siddharth-agrawal** [comment]
@concretevitamin Could you provide some hints as to how this might be done? I spent some time looking at some sparse operations to understand how `SparseTensor`s are handled. In most of the cases, `SparseTensor`s are converted to `Tensor`s for the underlying Op. One exception I found is `sparse_matmul`. Are `a_is_sparse`, `b_is_sparse` somehow relevant?

**Comment by concretevitamin** [comment]
@siddharth-agrawal Essentially, SparseTensor in Python can be thought of as a 3-tuple of dense tensors, `(indices, values, shape)`.  You're right that most underlying Ops take these three dense Tensors separately, but there's no "conversion" going on. 

As for `sparse_matmul()`: due to historical reasons it's actually a misnomer.  It does not operate on `SparseTensor`s, but instead use an optimized algorithm depending on the two flags.  So that op is irrelevant.

**Comment by siddharth-agrawal** [comment]
@concretevitamin If we take the special case of `tf.rank()`, then we need some way making [this statement](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/shape_ops.cc#L132) generic so that it works for a `SparseTensor` as well. Are there any utilities that can differentiate between a `Tensor` and a `SparseTensor`? If not, can you tell me how to approach this (since you said that i should be easy in the description)? I did a lot of digging, but I still have no idea how to solve this. Even a 'where to look' reply would be fine.

**Comment by concretevitamin** [comment]
@siddharth-agrawal I think this can be implemented entirely in Python.  In [array_ops.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/array_ops.py), you could add a `def rank(...)` function with the correct signature and convention.  Inside that, if `isinstance(t, ops.SparseTensor)`, you can return `size(t.shape)`, because the number of elements in `t.shape` is the rank of `t`. 

(More explanation: right now, when `tf.rank()` is called, what really happens is [this line](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/array_ops.py#L89). So once we override it properly, the Python frontend can handle the dispatch. )

**Comment by siddharth-agrawal** [comment]
@concretevitamin Is there any example of overriding? I'm not familiar with how it's done in TensorFlow. I saw the binary op overrides in `math_ops.py`, but I'm guessing this will be somewhat different.

**Comment by concretevitamin** [comment]
@siddharth-agrawal One recent example is the introduction of `tf.sparse_softmax()` [here](https://github.com/tensorflow/tensorflow/commit/7e5261b7f35871282010f655de5d72e2d2edc29f).  The Python wrapper in `sparse_ops.py` -- named `sparse_softmax()` and which accepts one input -- overrides the function `gen_sparse_ops.sparse_softmax()`, which accepts three inputs.  This is what I meant by overriding.  Does this make sense?

**Comment by siddharth-agrawal** [comment]
@concretevitamin Yes, this helps. I will try to get something out by Friday.

**Comment by siddharth-agrawal** [comment]
@concretevitamin Would like your comments on the PR. I retained the comment for `_sparsify()' so that you don't miss it while factoring things out.

**Comment by siddharth-agrawal** [comment]
@concretevitamin New PR for tf.shape(). Would like your comments.

**Comment by siddharth-agrawal** [comment]
@concretevitamin For `tf.size()`, what should the output be for `SparseTensor`? The size of the tensor or the number of non-zero values?

**Comment by concretevitamin** [comment]
I am leaning towards making it consistent with dense Tensors, i.e. it
returns the size of the tensor.  Consistency here might be the less
surprising choice.

(Also, quick heads up I will be out this coming week.  @ebrevdo and
@rmlarsen might be good persons to look at SparseTensor-related changes!)

On Sat, Jun 4, 2016 at 11:08 PM, Siddharth Agrawal <notifications@github.com

**Comment by ebrevdo** [comment]
+1 for having it return reduce_prod(sparse_tensor.shape)

**Comment by siddharth-agrawal** [comment]
@ebrevdo New PR for tf.size(). Would like your comments on it.

**Comment by girving** [comment]
Drive by comment to say thank you @siddharth-agrawal for all the PRs!

**Comment by siddharth-agrawal** [comment]
@girving My pleasure! It is fun working with the TF team!

**Comment by concretevitamin** [comment]
Closing this as all three methods are now checked in.  Thanks!

**Comment by siddharth-agrawal** [comment]
@concretevitamin Your comment [here](https://github.com/tensorflow/tensorflow/pull/2681#discussion_r66742242) applies to my earlier PR [here](https://github.com/tensorflow/tensorflow/pull/2659) too. I will update it once the `tf.size()` PR is merged.

**Comment by siddharth-agrawal** [comment]
@concretevitamin You can close the issue now, all changes done.

**Comment by concretevitamin** [comment]
@siddharth-agrawal Thanks for your contributions!

**Comment by siddharth-agrawal** [comment]
@concretevitamin My pleasure!
